src/shortest_path/mod.rs

```diff
@@ -1089,3 +1091,117 @@ pub fn graph_distance_matrix(
+///
+/// :returns: The average shortest path length. If the graph is empty this
+///     will return NaN, if there is a single node 0 will be returned and if
+///     the graph is disconnected it will return infinity.
+/// :rtype: float
+#[pyfunction(parallel_threshold = "300", as_undirected = "false")]
+#[pyo3(
+    text_signature = "(graph, /, parallel_threshold=300, as_undirected=False)"
+)]
+pub fn digraph_unweighted_average_shortest_path_length(
+    graph: &digraph::PyDiGraph,
+    parallel_threshold: usize,
+    as_undirected: bool,
+) -> f64 {
+    let n = graph.node_count() as f64;
+    if n == 0.0 {
+        return std::f64::NAN;
+    }
+    if n == 1.0 {
+        return 0.0;
+    }
+    if !is_connected(&graph.graph) {
+        return std::f64::INFINITY;
+    }
+    let sum = average_length::compute_distance_sum(
+        &graph.graph,
+        parallel_threshold,
+        as_undirected,
+    ) as f64;
+    sum / (n * (n - 1.0))
+}
+
+/// Return the average shortest path length for a :class:`~retworkx.PyGraph`
+/// with unweighted edges.
+///
+/// The average shortest path length is calculated as
+///
+/// .. math::
+///
+///     a =\sum_{s,t \in V} \frac{d(s, t)}{n(n-1)}
+///
+/// where :math:`V` is the set of nodes in ``graph``, :math:`d(s, t)` is the
+/// shortest path length from :math:`s` to :math:`t`, and :math:`n` is the
+/// number of nodes in ``graph``.
+///
+/// This function is also multithreaded and will run in parallel if the number
+/// of nodes in the graph is above the value of ``parallel_threshold`` (it
+/// defaults to 300). If the function will be running in parallel the env var
+/// ``RAYON_NUM_THREADS`` can be used to adjust how many threads will be used.
+/// By default it will use all available CPUs if the environment variable is
+/// not specified.
+///
+/// :param PyGraph graph: The graph to compute the average shortest path length
+///     for
+/// :param int parallel_threshold: The number of nodes to calculate the
+///     the distance matrix in parallel at. It defaults to 300, but this can
+///     be tuned to any number of nodes.
+///
+/// :returns: The average shortest path length. If the graph is empty this
+///     will return NaN, if there is a single node 0 will be returned and if
+///     the graph is disconnected it will return infinity.
+/// :rtype: float
+#[pyfunction(parallel_threshold = "300")]
+#[pyo3(text_signature = "(graph, /, parallel_threshold=300)")]
+pub fn graph_unweighted_average_shortest_path_length(
+    graph: &graph::PyGraph,
+    parallel_threshold: usize,
+) -> f64 {
+    let n = graph.node_count() as f64;
+    if n == 0.0 {
+        return std::f64::NAN;
+    }
+    if n == 1.0 {
+        return 0.0;
+    }
+    if !is_connected(&graph.graph) {
+        return std::f64::INFINITY;
+    }
+    let sum = average_length::compute_distance_sum(
+        &graph.graph,
+        parallel_threshold,
+        true,
+    ) as f64;
+    sum / (n * (n - 1.0))
```

Comment:
It looks a bit cleaner to me to have `let n = graph.node_count();`, i.e. `n` is an integer. Then something like `(n * (n - 1)) as f64`. That is, you still need only a single `as f64`, but you put it in the place where you really need it, rather than early, so that you need things like `n == 1.0` above.

Reply:
Good idea, done in: https://github.com/Qiskit/retworkx/pull/403/commits/fbe1fa55db9401188a95bb93940eda5be0bbc500